genesum
=======

This is a tool to create gene-level summaries of transcript expression
estimates.  You provide the tool with an annotation file (in GTF/GFF format)
and a set of transcript-level estimates, and it aggregates these expression
estimates to the gene-level.

example usage
-------------

Say you have a file `annotations.gtf` and a set of expression estimates
`expressions.sf` (e.g. generated by
[Sailfish](https://github.com/kingsfordgroup/sailfish)). This tool can be
invoked as such:

```
$ genesum -e expressions.sf -g annotations.gtf -o expressions_genes.sf
```

This will produce a file, `expressions_genes.sf` where the expression estimates
from `expressions.sf` have been aggregated to the gene level according to the
transcript-to-gene mapping encoded by `annotations.gtf`.  For simplicity, the
length assigned to each gene in the output file is simply the length of the
longest transcript present in the input file that mapped to that gene.
Currently, transcripts are grouped together based on the `gene_name` field of
the `gtf` file. However, the tool will soon support aggregating transcripts
based on other fields like `gene_id` or `locus_id`.

